File: examples/cpp/runner.cpp
#include <iostream>
#include <iomanip>
#include <unistd.h>

#include "pins.hpp"
#include "runner.hpp"
#include "pin_utils.hpp"
#include "configuration.hpp"

using std::cout;
using std::hex;
using std::dec;
using std::endl;
using std::right;
using std::setw;
using std::setfill;

Runner::Runner(int device, Memory *mem)
{
    this->device = device;
    this->mem = mem;
    this->phase = false;
    this->cycle = 0;
}

/**
 * CPU reset procedure. It requires to keep the CPU reset pin (40)
 * low for two cycles. After that this pin is set high.
 */
void Runner::reset()
{
    uint8_t b[] = {
        0b00001010, // BE, RW high
        0b00000000,
        0b00000000,
        0b00000000,
        0b10101000  // IRQ, NMI, VDD high
    };

    for (int i = 0; i < 4; i++) {
        this->pins = W65C02Pins(b);
        this->pins.set_overflow = true;
        this->pins.phi2 = this->phase;
        this->write_serial();
        usleep(CYCLE_DURATION);
        this->read_serial();
        this->advance_cycles();
    }

    this->pins.data = 0;
    this->pins.addr = 0;
    this->pins.reset = true;
    this->pins.ready = true;
    this->pins.set_overflow = true;
}

bool Runner::step()
{
    static uint16_t addr = 0;
    static bool write = false;

    this->pins.phi2 = this->phase;
    this->write_serial();
    usleep(CYCLE_DURATION);

    this->read_serial();
    if (!this->phase) {
        addr = this->pins.addr;
        write = this->pins.is_write();
        if (!write) {
            this->pins.data = this->mem->read_byte(addr);
        }
    } else {
        uint8_t data = this->pins.data;
        if (write) {
            this->mem->write_byte(addr, data);
        }
        this->print_state();
        if (EXIT_ON_BRK && data == 0 && this->pins.sync) {
            return false;
        }
    }

    this->advance_cycles();
    return MAX_CYCLES == 0 || this->cycle < MAX_CYCLES;
}

void Runner::run()
{
    this->reset();
    while (this->step());
}

void Runner::print_state()
{
    cout << "Cycle " << setfill(' ') << setw(5) << right << dec << this->cycle
         << ", Addr: " << setfill('0') << setw(4) << right << hex << this->pins.addr
         << " (" << addr_to_binary_str(this->pins.addr)
         << "), Data: " << setfill('0') << setw(2) << right << hex << (uint16_t)this->pins.data
         << ", R/W: " << (this->pins.rw ? "R" : "W")
         << ", VP: " << (int)this->pins.vector_pull
         << ", SYNC " << (int)this->pins.sync
         << ", IRQ " << (int)this->pins.irq
         << ", NMI " << (int)this->pins.nmi
         << ", RES " << (int)this->pins.reset
         << ", PHI1O " << (int)this->pins.phi1o
         << ", PHI2O " << (int)this->pins.phi2o
         << endl;
}

void Runner::read_serial()
{
    uint8_t buff[5];
    for (int n=0; !n;) {
        n = read(this->device, buff, BUFFSIZE);
    }
    this->pins.set_pins(buff);
}

void Runner::write_serial()
{
    uint8_t buff[5];
    this->pins.set_buff(buff);
    auto _ = write(this->device, buff, BUFFSIZE);
}

void Runner::advance_cycles()
{
    if (this->phase) {
        ++this->cycle;
    }
    this->phase = !this->phase;
}

File: examples/rust/src/runner.rs
use crate::{
    configuration::{CYCLE_DURATION, SHOW_RAW_DATA},
    pins::Pins,
};
use serialport::SerialPort;
use std::{fs::File, io::{self, Read}, path::PathBuf, thread::sleep};

pub struct Runner {
    pub cycle: u64,
    pub mem: [u8; 1 << 16],
    pub port: Box<dyn SerialPort>,
    pub phase: bool,
    pub addr: u16,
    pub write: bool,
    pub pins: Pins,
}

impl Runner {
    /// It executes a single half-step (high or low clock signal) of the CPU.
    /// For every exection it sends data to CPU, adjusting the clock status (PHI2) first,
    /// and then it reads status back from the CPU.
    /// For every low half-cycle it reads the address and the RWB status (pin 34).
    /// For every high half-cycle it does the actual memory operation.
    /// At the end the method advances the clock (switches the phase and increments the
    /// clock counter).
    /// The method returns boolean value indicating whether the program reached an end,
    /// that is assumed to be represented by BRK instruction (opcode 0).
    pub fn step(&mut self) -> bool {
        self.pins.phi2 = self.phase;
        self.write_port();

        sleep(CYCLE_DURATION);
        self.read_port();

        if !self.phase {
            self.addr = self.pins.addr;
            self.write = self.pins.is_write();
            if !self.write {
                self.pins.data = read_byte(&self.mem, self.addr);
            }
        } else {
            if self.write {
                write_byte(&mut self.mem, self.addr, self.pins.data);
            }
            println!("[{}] {}", self.cycle, self.pins);
            if self.pins.sync && self.pins.data == 0 && self.cycle > 20 {
                return false;
            }
        }

        self.advance_cycles();
        true
    }

    /// Resets the CPU, by holding RESB signal (pin 40) low for two cycles (4 half-cycles).
    pub fn reset(&mut self) {
        let mut pins = Pins::default();
        pins.be = true;
        pins.rw = true;
        pins.irq = true;
        pins.nmi = true;
        pins.so = false;

        for _ in 0..4 {
            self.pins = pins.clone(); // Pins::from(pins);
            self.pins.phi2 = self.phase;
            self.write_port();

            sleep(CYCLE_DURATION);
            self.read_port();
            self.advance_cycles();
        }

        self.pins.data = 0;
        self.pins.addr = 0;
        self.pins.reset = true;
        self.pins.ready = true;
        self.pins.so = true;
    }

    /// Reads 5-bytes buffer from the serial port and updates
    /// `pins` field.
    fn read_port(&mut self) {
        let mut buff: [u8; 5] = [0; 5];
        self.port
            .read(&mut buff)
            .expect("Read error from serial port");
        if SHOW_RAW_DATA {
            print!("Reading: ");
            print_buff(&buff);
        }
        self.pins = Pins::from(buff);
    }

    /// Send the value of `pins` field into to serial port.
    fn write_port(&mut self) {
        let buff: [u8; 5] = self.pins.into();
        if SHOW_RAW_DATA {
            print!("Writing: ");
            print_buff(&buff);
        }
        self.port.write(&buff).expect("Write error to serial port");
    }

    /// Changes phase of the clock and advances the clock count.
    fn advance_cycles(&mut self) {
        if self.phase {
            self.cycle += 1;
        }
        self.phase = !self.phase;
    }
}

/// Writes a byte under a given address to memory emulation.
pub fn write_byte(mem: &mut [u8], addr: u16, val: u8) {
    mem[addr as usize] = val;
}

/// Reads a byte from a given address of memeory emulation.
pub fn read_byte(mem: &[u8], addr: u16) -> u8 {
    mem[addr as usize]
}

/// Reads contnet of a 6502 program and returns it as
/// a vector of bytes.
fn get_file_as_byte_vec(filename: &PathBuf) -> io::Result<Vec<u8>> {
    let mut f = File::open(filename)?;
    let mut buffer = Vec::new();
    f.read_to_end(&mut buffer)?;
    Ok(buffer)
}

/// Loads 6502 program into memory emulation under a given address.
pub fn load_program(mem: &mut [u8], file: &str, addr: u16) {
    let prg = get_file_as_byte_vec(&PathBuf::from(file)).expect("Can't open program file");
    for (i, byte) in prg.iter().enumerate() {
        write_byte(mem, addr + (i as u16), *byte);
    }
}

/// Prints serial port buffer in a form of 4x8 bits.
fn print_buff(buff: &[u8]) {
    for i in 0..buff.len() {
        print!("{:08b} ", buff[i]);
    }
    println!();
}
